src/Model/common.hpp

```diff
@@ -114,65 +122,121 @@ void seissol::model::getTransposedElasticCoefficientMatrix( seissol::model::Elas
+  std::array<std::array<int, 2>, 3> traction_indices = {{{0,0}, {1,3}, {2,5}}};
+  std::array<std::array<int, 2>, 3> velocity_indices = {{{0,6}, {1,7}, {2,8}}};
+  //Eigen does not have fancy slicing options :(
+  using Matrix33 = Eigen::Matrix<real, 3, 3>;
+  Matrix33 R11 = Matrix33::Zero(); 
+  for (auto &t: traction_indices) {
+    for (int i = 0; i < 3; i++) {
+      R11(t[0], i) = R(t[1],i);
+    }
+  }
+  Matrix33 R21 = Matrix33::Zero();
+  for (auto &v: velocity_indices) {
+    for (int i = 0; i < 3; i++) {
+      R21(v[0], i) = R(v[1],i);
+    }
+  }
+  auto S = - (R21 * R11.inverse()).eval();
+
+  //set lower left block
+  for (auto &t: traction_indices) {
+    for (auto &v: velocity_indices) {
+      QgodLocal(t[1], v[1]) = S(v[0],t[0]);
+    }
+  }
+  //set lower right block
+  for (auto &v : velocity_indices) {
+    QgodLocal(v[1], v[1]) = 1.0;
+  }
 }
 
-template<typename T>
-void seissol::model::applyBoundaryConditionToElasticFluxSolver( enum ::faceType type,
-                                                                T&              QgodNeighbor )
+template<typename Tloc, typename Tneigh>
+void seissol::model::getTransposedElasticGodunovState( Material const&                      local,
+    Material const&                      neighbor,
+    enum ::faceType                      faceType,
+    Tloc&                                QgodLocal,
+    Tneigh&                              QgodNeighbor )
 {
-  if (type == freeSurface) {
-    // Gamma is a diagonal matrix
-    real Gamma[] = { -1.0, 1.0, 1.0, -1.0, 1.0, -1.0, 1.0, 1.0, 1.0 };
-    // Gamma^T * Fneighbor
-    for (unsigned j = 0; j < QgodNeighbor.shape(1); ++j) {
-      for (unsigned i = 0; i < 9; ++i) {
-        QgodNeighbor(i,j) *= Gamma[i];
+  QgodNeighbor.setZero();
+
+  Matrix99 R = Matrix99::Zero();
+
+  //eigenvectors have been precalculated
+  R(0,0) = local.lambda + 2*local.mu;
+  R(0,8) = neighbor.lambda + 2*neighbor.mu;
+  R(1,0) = local.lambda;
+  R(1,4) = 1;
+  R(1,8) = neighbor.lambda;
+  R(2,0) = local.lambda;
+  R(2,5) = 1;
+  R(2,8) = neighbor.lambda;
+  R(3,1) = local.mu;
+  R(3,7) = neighbor.mu;
+  R(4,3) = 1;
+  R(5,2) = local.mu;
+  R(5,6) = neighbor.mu;
+  R(6,0) = sqrt((local.lambda + 2*local.mu)/local.rho);
+  R(6,8) = -sqrt((neighbor.lambda + 2*neighbor.mu)/neighbor.rho);
+  R(7,1) = sqrt(local.mu/local.rho);
+  R(7,7) = -sqrt(neighbor.mu/neighbor.rho);
+  R(8,2) = sqrt(local.mu/local.rho);
+  R(8,6) = -sqrt(neighbor.mu/neighbor.rho);
+
+  if(faceType == freeSurface) {
+    getTransposedBoundaryGodunovState(QgodLocal, QgodNeighbor, R);
+
+  } else {
+    Matrix99 R_inv = Matrix99::Zero();
+    R_inv(0,0) = 1/(local.lambda + 2*local.mu) - (neighbor.lambda + 2*neighbor.mu)*sqrt((local.lambda + 2*local.mu)/local.rho)/((local.lambda + 2*local.mu)*(local.lambda + 2*local.mu)*((neighbor.lambda + 2*neighbor.mu)*sqrt((local.lambda + 2*local.mu)/local.rho)/(local.lambda + 2*local.mu) + sqrt((neighbor.lambda + 2*neighbor.mu)/neighbor.rho)));
```

Comment:
Is this generated code? I'm just wondering whether it would be more readable if one codes Rinv by inverting the three independent 2x2 matrices.